interpreter/src/main/scala/pl/touk/nussknacker/engine/dict/ProcessDictSubstitutor.scala

```diff
@@ -19,44 +20,89 @@ import pl.touk.nussknacker.engine.spel.ast.{
 
 class ProcessDictSubstitutor(
     dictRegistry: DictRegistry,
     replacingStrategy: ReplacingStrategy,
-    prepareSubstitutionsCollector: (ExpressionTypingInfo, ReplacingStrategy) => Option[ExpressionSubstitutionsCollector]
+    prepareSubstitutionsCollector: (
+        ExpressionTypingInfo,
+        ReplacingStrategy
+    ) => Option[ExpressionSubstitutionsCollector],
+    isReverse: Boolean
 ) extends LazyLogging {
 
   def substitute(
       process: CanonicalProcess,
       processTypingInfo: Map[String, Map[String, ExpressionTypingInfo]]
   ): CanonicalProcess = {
+
     val rewriter = ProcessNodesRewriter.rewritingAllExpressions { exprIdWithMetadata => expr =>
       val nodeExpressionId             = exprIdWithMetadata.expressionId
       val nodeTypingInfo               = processTypingInfo.getOrElse(nodeExpressionId.nodeId.id, Map.empty)
       val optionalExpressionTypingInfo = nodeTypingInfo.get(nodeExpressionId.expressionId)
-      val substitutedExpression = optionalExpressionTypingInfo
-        .flatMap(prepareSubstitutionsCollector(_, replacingStrategy))
-        .map { substitutionsCollector =>
-          val substitutions     = substitutionsCollector.collectSubstitutions(expr)
-          val afterSubstitution = ExpressionSubstitutor.substitute(expr.expression, substitutions)
-          if (substitutions.nonEmpty)
-            logger.debug(
-              s"Found ${substitutions.size} substitutions in expression: ${process.name} > ${nodeExpressionId.nodeId.id} > ${nodeExpressionId.expressionId}. " +
-                s"Expression: '${expr.expression}' replaced with '$afterSubstitution'"
-            )
-          afterSubstitution
-        }
-        .getOrElse(expr.expression)
-      expr.copy(expression = substitutedExpression)
+
+      if (expr.language == Expression.Language.DictKeyWithLabel) {
+        if (!isReverse)
+          removeLabelFromDictKeyExpression(expr) // no need to keep label in BE
+        else
+          addLabelToDictKeyExpression(process, optionalExpressionTypingInfo, nodeExpressionId)
+      } else
+        substituteExpression(process, expr, optionalExpressionTypingInfo, nodeExpressionId)
     }
 
     rewriter.rewriteProcess(process)
   }
 
+  private def removeLabelFromDictKeyExpression(expr: Expression) =
+    DictKeyWithLabelExpressionParser
+      .parseDictKeyWithLabelExpression(expr.expression)
+      .map(expr => Expression.dictKeyWithLabel(expr.key, None))
+      .leftMap(e => throw new IllegalStateException(s"Errors parsing DictKeyWithLabel expression: $expr: $e"))
+      .toOption
+      .get
+
+  private def addLabelToDictKeyExpression(
+      process: CanonicalProcess,
+      optionalExpressionTypingInfo: Option[ExpressionTypingInfo],
+      nodeExpressionId: NodeExpressionId
+  ) =
+    optionalExpressionTypingInfo match {
+      case Some(DictKeyWithLabelExpressionTypingInfo(key, label, _)) => Expression.dictKeyWithLabel(key, label)
+      case _ =>
```

Comment:
I'm not sure if we should handle `None` the same way. Substition is done even if you got validation error on the compilation step. I think that we should return something like missing label that user could fix on the FE side by choosing another label. WDYT?